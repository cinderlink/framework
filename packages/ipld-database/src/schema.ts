import { DIDDagInterface } from "@candor/core-types";
import Emittery from "emittery";
import type { CID } from "multiformats";
import { Table, TableDefinition, TableRow } from "./table";

export type SavedSchema = {
  name: string;
  defs: Record<string, TableDefinition>;
  tables: Record<string, string | undefined>;
};

export type SchemaEvents = {
  "/schema/loaded": undefined;
};

export class Schema extends Emittery<SchemaEvents> {
  public tables: Record<string, Table> = {};
  constructor(
    public name: string,
    public defs: Record<string, TableDefinition>,
    public dag: DIDDagInterface,
    public encrypted = true
  ) {
    super();
    Object.entries(defs).forEach(([name, def]) => {
      this.tables[name] = new Table(def, this.dag);
    });
  }

  async createTable(name: string, def: TableDefinition) {
    this.tables[name] = new Table(def, this.dag);
  }

  async dropTable(name: string) {
    delete this.tables[name];
  }

  getTable<T extends TableRow = TableRow>(name: string) {
    return this.tables[name] as Table<T>;
  }

  async save() {
    console.info("saving schema", this.name);
    const tables: Record<string, string | undefined> = {};
    await Promise.all(
      Object.entries(this.tables).map(async ([name]) => {
        const tableCID = await this.tables[name].save();
        if (tableCID) {
          tables[name] = tableCID?.toString();
        }
      })
    );
    console.info("storing schema", this.name, this.defs, tables);
    const savedSchema = {
      name: this.name,
      defs: this.defs,
      tables,
    };
    return this.encrypted
      ? this.dag.storeEncrypted(savedSchema)
      : this.dag.store(savedSchema);
  }

  static async load(cid: string | CID, dag: DIDDagInterface, encrypted = true) {
    const data = encrypted
      ? await dag.loadDecrypted<SavedSchema>(cid)
      : await dag.load<SavedSchema>(cid);
    return Schema.fromSavedSchema(data, dag);
  }

  static async fromSavedSchema(
    data: SavedSchema,
    dag: DIDDagInterface,
    encrypted = true
  ) {
    if (!data) throw new Error("Invalid schema data");
    console.info("loading from saved schema", data);
    const schema = new Schema(data.name, data.defs, dag, encrypted);
    await Promise.all(
      Object.entries(data.tables).map(async ([name, tableCID]) => {
        console.info("loading table", { name, tableCID });
        await schema.tables[name]?.load(tableCID!);
      })
    );
    schema.emit("/schema/loaded");
    return schema;
  }
}
