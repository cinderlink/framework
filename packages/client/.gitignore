candor/
.turbo/
dist/
node_modules/