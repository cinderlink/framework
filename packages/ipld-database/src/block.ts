import Minisearch from "minisearch";
import {
  BlockAggregates,
  BlockData,
  BlockFilters,
  BlockHeaders,
  TableBlockInterface,
  TableDefinition,
  TableInterface,
  TableRow,
} from "@candor/core-types";
import type { CID } from "multiformats/cid";

import * as json from "multiformats/codecs/json";
import { base58btc } from "multiformats/bases/base58";

export class TableBlock<
  Row extends TableRow = TableRow,
  Def extends TableDefinition = TableDefinition
> implements TableBlockInterface<Row, Def>
{
  public index: Minisearch;
  private _changed: boolean = false;

  constructor(
    public table: TableInterface<Row, Def>,
    public cid: CID | undefined,
    public cache: Partial<BlockData<Row>> = {}
  ) {
    this.index = this.cache.filters?.search
      ? Minisearch.loadJS(
          this.cache.filters.search,
          this.table.def.searchOptions
        )
      : new Minisearch(this.table.def.searchOptions);
  }

  get changed() {
    return this._changed;
  }

  async prevCID(): Promise<string | undefined> {
    if (this.cache.prevCID) {
      return this.cache.prevCID;
    }

    if (!this.cid) {
      return undefined;
    }

    this.cache.prevCID = await this.table.dag.load<string>(
      this.cid,
      "/prevCID"
    );

    return this.cache.prevCID;
  }

  async getCID(): Promise<CID | undefined> {
    return this.cid ? this.cid : this.save();
  }

  async headers() {
    if (!this.cache.headers && !this.cid) {
      this.cache.headers = {
        schema: this.table.def.schema.id,
        table: this.table.tableId,
        encrypted: this.table.def.encrypted,
        index: 0,
        recordsFrom: 1,
        recordsTo: 0,
      } as BlockHeaders;
    } else if (this.cid) {
      this.cache.headers = await this.table.dag.load<BlockHeaders>(
        this.cid!,
        "/headers"
      );
    }

    if (!this.cache.headers) {
      throw new Error(`Block headers not found: ${this.cid}/headers`);
    }

    return this.cache.headers;
  }

  async filters() {
    if (!this.cache.filters && !this.cid) {
      this.cache.filters = {
        indexes: {},
        aggregates: {},
      } as BlockFilters;
    } else if (!this.cache.filters) {
      this.cache.filters = await this.table.dag.load<BlockFilters>(
        this.cid!,
        "/filters"
      );
    }

    if (!this.cache.filters) {
      throw new Error(`Block filters not found: ${this.cid}/filters`);
    }

    return this.cache.filters;
  }

  async records() {
    if (!this.cache.records && !this.cid) {
      this.cache.records = {} as Record<number, Row>;
    } else if (!this.cache.records) {
      this.cache.records = await this.table.dag.load<Row[]>(
        this.cid!,
        "/records"
      );
    }

    if (!this.cache.records) {
      throw new Error(`Block records not found: ${this.cid}/records`);
    }

    return this.cache.records;
  }

  async recordById(id: number) {
    if (this.cache.records) {
      return this.cache.records[id];
    }

    return this.table.dag.load<Row>(this.cid!, `/records/${id}`);
  }

  async hasIndex(name: string) {
    const indexes = (await this.filters()).indexes;
    return indexes[name] !== undefined;
  }

  async getIndex(name: string) {
    return (await this.filters()).indexes[name];
  }

  /**
   * Determine if a value is indexed in this block
   * @param name
   * @param value
   * @returns
   */
  async hasIndexValue(name: string, value: string, key: number = 0) {
    const indexes = (await this.filters()).indexes;
    return (
      indexes[name] &&
      Object.values(indexes[name]).some((index) => {
        return key === undefined
          ? index.values.some((v) => v == value)
          : index.values[key] == value;
      })
    );
  }

  /**
   * Determine if a value is indexed in this block
   * @param name
   * @param value
   * @returns
   */
  async hasIndexValues(name: string, value: string[], id?: number) {
    const indexes = (await this.filters()).indexes;
    const serialized = TableBlock.serialize(value);
    return id
      ? indexes[name]?.[serialized]?.ids.includes(id)
      : indexes[name]?.[serialized] !== undefined;
  }

  /**
   * Remove an index from the block
   * @param index
   * @param value
   * @param id
   */
  async removeIndexWithValues(index: string, values: string[], id: number) {
    const filters = await this.filters();
    const serialized = TableBlock.serialize(values);
    if (filters.indexes[index]?.[serialized]?.ids.includes(id)) {
      filters.indexes[index][serialized].ids.splice(
        filters.indexes[index][serialized].ids.indexOf(id),
        1
      );
      if (!filters.indexes[index][serialized].ids.length) {
        delete filters.indexes[index][serialized];
      }
    }

    this.cache.filters = filters;
    this._changed = true;
  }

  /**
   * Index values in the block
   * @param index
   * @param value
   * @param id
   */
  async addIndexWithValues(index: string, values: string[], id: number) {
    if (!this.table.def.indexes[index]) {
      throw new Error(`Index ${index} not found`);
    }
    const filters = await this.filters();
    if (!filters.indexes) {
      filters.indexes = {};
    }
    if (!filters.indexes[index]) {
      filters.indexes[index] = {};
    }
    const serialized = TableBlock.serialize(values);
    if (!filters.indexes[index][serialized]) {
      filters.indexes[index][serialized] = {
        values,
        ids: [id],
      };
    } else {
      if (
        this.table.def.indexes[index].unique &&
        filters.indexes[index][serialized]?.ids.length &&
        filters.indexes[index][serialized].ids[0] !== id
      ) {
        throw new Error(
          `Unique index violation: ${index} already has value ${values}`
        );
      }
      if (!filters.indexes[index][serialized].ids.includes(id)) {
        filters.indexes[index][serialized].ids.push(id);
      }
    }

    this.cache.filters = filters;
    this._changed = true;
  }

  /**
   * Get the ids of records that match the indexed value
   * @param name
   * @param value
   * @returns
   */
  async getIndexMatchIds(name: string, value: string) {
    if (!(await this.hasIndexValue(name, value))) return [];
    const indexes = (await this.filters()).indexes;
    return Object.values(indexes[name])
      .filter((index) => index.values.includes(value))
      .map((index) => index.ids)
      .flat();
  }

  async aggregate() {
    const filters = await this.filters();
    const records = await this.records();
    const entries = Object.entries(records);
    filters.aggregates = {} as BlockAggregates;

    Object.entries(this.table.def.aggregate).forEach(([key, type]) => {
      if (type === "sum") {
        filters.aggregates[key] = entries.reduce(
          (acc, [, row]) => acc + (row[key] as number),
          0
        );
      } else if (type === "count") {
        filters.aggregates[key] = entries.reduce(
          (acc, [, row]) => acc + (row[key] ? 1 : 0),
          0
        );
      } else if (type === "avg") {
        filters.aggregates[key] = entries.reduce(
          (acc, [, row]) => acc + (row[key] as number),
          0
        );
      } else if (type === "min") {
        filters.aggregates[key] = Math.min(
          ...entries.map(([, row]) => row[key] as number)
        );
      } else if (type === "max") {
        filters.aggregates[key] = Math.max(
          ...entries.map(([, row]) => row[key] as number)
        );
      } else if (type === "distinct") {
        filters.aggregates[key] = [
          ...new Set(entries.map(([, row]) => row[key] as string)),
        ];
      } else if (type === "range") {
        filters.aggregates[key] = [
          Math.min(...entries.map(([, row]) => row[key] as number)),
          Math.max(...entries.map(([, row]) => row[key] as number)),
        ];
      }
    });

    this.cache.filters = filters;
    this._changed = true;
    return filters.aggregates;
  }

  async violatesUniqueConstraints(row: Omit<Row, "id">, id?: number) {
    for (const [name, index] of Object.entries(this.table.def.indexes)) {
      const values = index.fields.map((f) => row[f] as string);
      if (index.unique && (await this.hasIndexValues(name, values, id))) {
        return true;
      }
    }
    return false;
  }

  async assertUniqueConstraints(row: Omit<Row, "id">, id?: number) {
    for (const [name, index] of Object.entries(this.table.def.indexes)) {
      const values = index.fields.map((f) => row[f] as string);
      if (index.unique && (await this.hasIndexValues(name, values, id))) {
        throw new Error(
          `ipld-database/block: Unique index violation: index ${name} already has values: ${values}`
        );
      }
    }
  }

  async addRecord(row: Row) {
    await this.assertUniqueConstraints(row, row.id);
    if (!this.cache.headers) {
      throw new Error(
        `ipld-database/block: Cannot set record ${row.id} without headers`
      );
    }

    if (row.id > this.cache.headers!.recordsTo + 1) {
      throw new Error(
        `ipld-database/block: Record id ${row.id} is too high for this block`
      );
    }
    this.cache.headers!.recordsTo = row.id;

    // add necessary indexes
    const indexes = this.table.def.indexes;
    for (const [name, index] of Object.entries(indexes)) {
      await this.addIndexWithValues(
        name,
        index.fields.map((f) => row[f] as string),
        row.id
      );
    }

    if (!this.cache.records) {
      this.cache.records = {};
    }

    this.cache.records[row.id] = row;
    this._changed = true;
  }

  async updateRecord(id: number, update: Partial<Row>) {
    await this.assertUniqueConstraints(update as Row, id);

    const records = await this.records();
    records[id] = { ...records[id], ...update };
    this.cache.records = records;
    this._changed = true;
  }

  async deleteRecord(id: number) {
    // delete from indexes
    const indexes = this.table.def.indexes;
    for (const [name, index] of Object.entries(indexes)) {
      await this.removeIndexWithValues(
        name,
        index.fields.map((f) => this.cache.records![id][f] as string),
        id
      );
    }
    delete this.cache!.records![id];
    this._changed = true;
  }

  async search(query: string): Promise<Row[]> {
    const results = this.index.search(query);
    const records: Row[] = [];
    for (const result of results) {
      const record = await this.recordById(result.id);
      if (record) {
        records.push(record);
      }
    }

    return records;
  }

  async load(force = false) {
    if (this._changed && !force) {
      throw new Error(
        "Block has unsaved changes, refusing to load without [force=true]"
      );
    }

    await this.headers();
    await this.filters();
    await this.records();
  }

  async save() {
    console.info(`ipld/block: block save requested`);
    if (!this._changed) {
      console.info("ipld/block: no changes to save");
      return this.cid;
    }

    const recordsFrom = Math.min(
      ...Object.keys(this.cache.records || {}).map(Number)
    );
    const recordsTo = Math.max(
      ...Object.keys(this.cache.records || {}).map(Number)
    );

    if (recordsFrom !== this.cache.headers!.recordsFrom) {
      console.info(`ipld/block: invalid block headers`);
      throw new Error(
        `ipld/block: recordsFrom mismatch: ${recordsFrom} (min id) !== ${
          this.cache.headers!.recordsFrom
        } (headers)`
      );
    }
    if (recordsTo !== this.cache.headers!.recordsTo) {
      console.info(`ipld/block: invalid block headers`);
      throw new Error(
        `ipld/block: recordsFrom mismatch: ${recordsTo} (max id) !== ${
          this.cache.headers!.recordsTo
        } (headers)`
      );
    }

    console.info(`ipld/block: aggregating block`);
    await this.aggregate();

    console.info(`ipld/block: storing block`);
    this.cid = this.table.encrypted
      ? await this.table.dag.storeEncrypted(
          this.toJSON() as Record<string, any>
        )
      : await this.table.dag.store(this.toJSON());

    console.info(
      `ipld/block: saved block (prev: ${this.cache.prevCID}, current: ${this.cid})`
    );

    return this.cid;
  }

  get needsRollup() {
    return (
      this.cache.headers!.recordsTo >=
      this.cache.headers!.recordsFrom + this.table.def.rollup
    );
  }

  toJSON() {
    if (!this.cache) {
      throw new Error("Block not loaded");
    }

    return this.cache as BlockData<Row>;
  }

  toString() {
    return TableBlock.serialize(this.toJSON());
  }

  static fromJSON<T extends TableRow = TableRow>(
    table: TableInterface<T>,
    cache: BlockData<T>,
    cid?: CID
  ) {
    return new TableBlock<T>(table, cid, cache);
  }

  static fromString<T extends TableRow = TableRow>(
    value: string,
    table: TableInterface,
    cid?: CID
  ) {
    return new TableBlock(
      table,
      cid,
      TableBlock.deserialize(value) as BlockData<T>
    );
  }

  static deserialize(value: string) {
    return json.decode(base58btc.decode(value));
  }

  static serialize(value: any) {
    return base58btc.encode(json.encode(value));
  }
}
