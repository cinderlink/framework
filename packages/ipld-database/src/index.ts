export * from "./schema";
export * from "./table";
