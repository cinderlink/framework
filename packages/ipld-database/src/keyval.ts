import { CID } from "multiformats/cid";
import { DagKeyval, DAGInterface } from "@candor/core-types";

export class IPLDKeyval<
  Data extends Record<string, unknown> = Record<string, unknown>
> implements DagKeyval
{
  public data: Data;
  public loaded: boolean = false;
  constructor(
    public id: string,
    public dag: DAGInterface,
    public encrypted = false
  ) {
    this.data = {} as Data;
  }

  async load(cid: CID) {
    const data = await this.dag.load<Data>(cid);
    if (!data) {
      throw new Error("Failed to load keyval store");
    }

    this.data = data;
    this.loaded = true;
  }
}
