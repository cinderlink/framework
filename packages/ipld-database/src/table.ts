import type { DIDDagInterface } from "@candor/core-types";
import { CID } from "multiformats";
import Emittery from "emittery";
import Ajv from "ajv";
import Minisearch from "minisearch";
import {
  TableBlock,
  TableDefinition,
  TableEvents,
  TableInterface,
  TableRow,
} from "@candor/core-types/src/database/table";

const ajv = new Ajv();

export class Table<T extends TableRow = TableRow>
  extends Emittery<TableEvents>
  implements TableInterface<T>
{
  currentIndex: number = 0;
  currentBlock: TableBlock;
  encrypted: boolean;

  constructor(public def: TableDefinition, public dag: DIDDagInterface) {
    super();
    this.encrypted = def.encrypted;
    this.currentBlock = this.createBlock();
  }

  createBlock(prevCID: string | undefined = this.currentBlock?.prevCID) {
    return {
      prevCID,
      fromIndex: this.currentIndex,
      toIndex: this.currentIndex + this.def.rollup,
      records: {},
      indexes: {},
      aggregates: {},
      search: new Minisearch(this.def.searchOptions),
    };
  }

  async insert(data: T) {
    this.assertValid(data);
    const rowCID = this.encrypted
      ? await this.dag.storeEncrypted(data)
      : await this.dag.store(data);
    if (!rowCID) {
      throw new Error("Failed to store row");
    }
    data.id = this.currentIndex++;
    this.currentBlock.records[data.id] = data;
    this.def.indexes.forEach((index) => {
      if (!this.currentBlock.indexes[index]) {
        this.currentBlock.indexes[index] = {};
      }
      const value = data[index] as string;
      if (!this.currentBlock.indexes[index][value]) {
        this.currentBlock.indexes[index][value] = [data.id as number];
      } else {
        this.currentBlock.indexes[index][value].push(data.id as number);
      }
    });
    this.currentBlock.search?.add(data);
    if (this.currentIndex - this.currentBlock.fromIndex >= this.def.rollup) {
      await this.rollup();
    }

    return data.id;
  }

  async search(query: string, limit = 10) {
    let results: T[] = [];
    await this.unwind(async (block) => {
      const searchResults = block.search?.search(query);
      const searchIds = [...new Set(searchResults?.map((r) => r.id) || [])];
      const records: T[] = searchIds.map((id) => block.records[id] as T);
      if (records) {
        results = results.concat(records);
      }
      return results.length < limit;
    });
    return results.slice(0, limit);
  }

  async rollup() {
    this.currentBlock = this.updateBlockAggregates(this.currentBlock);
    this.currentBlock.index = this.currentBlock.search?.toJSON();
    this.currentBlock.toIndex = this.currentIndex;
    delete this.currentBlock.search;
    if (!this.currentBlock.prevCID) {
      // can't save undefined
      delete this.currentBlock.prevCID;
    }
    const blockCID = this.encrypted
      ? await this.dag.storeEncrypted(this.currentBlock)
      : await this.dag.store(this.currentBlock);
    this.currentBlock = this.createBlock(blockCID?.toString());
  }

  updateBlockAggregates(block: TableBlock) {
    block.aggregates = {};
    const records = Object.values(block.records);
    Object.entries(this.def.aggregate).forEach(([key, type]) => {
      if (type === "sum") {
        block.aggregates[key] = records.reduce(
          (acc, row) => acc + (row[key] as number),
          0
        );
      } else if (type === "count") {
        block.aggregates[key] = records.reduce(
          (acc, row) => acc + (row[key] ? 1 : 0),
          0
        );
      } else if (type === "avg") {
        block.aggregates[key] = records.reduce(
          (acc, row) => acc + (row[key] as number),
          0
        );
      } else if (type === "min") {
        block.aggregates[key] = Math.min(
          ...records.map((row) => row[key] as number)
        );
      } else if (type === "max") {
        block.aggregates[key] = Math.max(
          ...records.map((row) => row[key] as number)
        );
      } else if (type === "distinct") {
        block.aggregates[key] = [
          ...new Set(records.map((row) => row[key] as string)),
        ];
      } else if (type === "range") {
        block.aggregates[key] = [
          Math.min(...records.map((row) => row[key] as number)),
          Math.max(...records.map((row) => row[key] as number)),
        ];
      }
    });
    return block;
  }

  async save() {
    if (Object.values(this.currentBlock.records).length > 0) {
      await this.rollup();
    }
    return this.currentBlock.prevCID;
  }

  async update(id: number, data: Partial<T>) {
    // we want to seek backwards from the current block
    // to find the block containing the row with the given id
    // we then want to update the row in that block
    // and then re-write that block and all subsequent blocks
    // to update the indexes, aggregates, and CID references
    // we also want to update the search index

    let blocksUnwound: string[] = [];
    let rewriteBlock: TableBlock | undefined = undefined;

    const unwindFn = async (block: TableBlock, cid: string | undefined) => {
      // if the block contains the row we want to update
      const row = block.records[id];
      const updated = { ...row, ...data };
      if (row) {
        block.records[id] = updated;
        // update the indexes
        this.def.indexes.forEach((index) => {
          // if the index has changed
          if (updated[index] && updated[index] !== row[index]) {
            // remove the row from the old index
            block.indexes[index][row[index] as string] = block.indexes[index][
              row[index] as string
            ].filter((rowID) => rowID !== id);
            // add the row to the new index
            if (!block.indexes[index][updated[index] as string]) {
              block.indexes[index][updated[index] as string] = [id];
            } else {
              block.indexes[index][updated[index] as string].push(id);
            }
          }
        });
        if (cid) {
          // update the aggregates
          block = this.updateBlockAggregates(block);
          // update the search index
          block.search?.removeAll();
          block.search?.addAll(Object.values(block.records));
          // update the CID reference
          rewriteBlock = block;
        }
        return false;
      }
      return true;
    };

    await this.unwind(unwindFn);

    // rewrite the blocks
    if (rewriteBlock) {
      let prevCID = this.encrypted
        ? await this.dag.storeEncrypted(rewriteBlock)
        : await this.dag.store(rewriteBlock);
      for (const cid of blocksUnwound) {
        const block = this.encrypted
          ? await this.dag.loadDecrypted<TableBlock>(cid)
          : await this.dag.load<TableBlock>(cid);
        if (block) {
          block.prevCID = prevCID?.toString();
          prevCID = this.encrypted
            ? await this.dag.storeEncrypted(block)
            : await this.dag.store(block);
        }
      }
    }

    return id;
  }

  async load(cid: CID | string) {
    const block = this.encrypted
      ? await this.dag.loadDecrypted<TableBlock>(cid)
      : await this.dag.load<TableBlock>(cid);
    if (block) {
      this.currentBlock = block;
      this.currentBlock.search = this.currentBlock.index
        ? await Minisearch.loadJS(
            this.currentBlock.index,
            this.def.searchOptions
          )
        : new Minisearch(this.def.searchOptions);
      this.currentIndex = block.toIndex;
      // this.currentBlock = this.createBlock(cid.toString());
    } else {
      this.currentBlock = this.createBlock();
    }
    this.emit("/table/loaded");
  }

  async unwind(
    test: (block: TableBlock, cid: string | undefined) => Promise<boolean>
  ) {
    let cid: string | undefined = undefined;
    let block: TableBlock | undefined = this.currentBlock;
    let prevented = false;
    while (prevented === false && block) {
      prevented = !(await test(block, cid).catch((err) => {
        console.error("unwind error", err);
        return false;
      }));
      if (prevented) {
        cid = undefined;
        block = undefined;
        return;
      }
      cid = block.prevCID;
      block = block.prevCID
        ? this.encrypted
          ? await this.dag.loadDecrypted<TableBlock>(block.prevCID)
          : await this.dag.load<TableBlock>(block.prevCID)
        : undefined;
    }
  }

  async findByIndex(index: string, value: string) {
    let result: T | undefined;
    await this.unwind(async (block) => {
      if (block.indexes[index]?.[value]) {
        for (let rowID of block.indexes[index][value]) {
          result = block.records[rowID] as T;
          if (result) {
            return false;
          }
        }
      }
      return true;
    });
    return result;
  }

  async where(match: (row: T) => boolean) {
    const results: T[] = [];
    await this.unwind(async (block) => {
      for (let row of Object.values(block.records)) {
        if (match(row as T)) {
          results.push(row as T);
        }
      }
      return true;
    });
    return results;
  }

  async find(match: (row: T) => boolean) {
    let result: T | undefined;
    await this.unwind(async (block) => {
      for (let row of Object.values(block.records)) {
        if (match(row as T)) {
          result = row as T;
          return false;
        }
      }
      return true;
    });
    return result;
  }

  async upsert(index: string, value: string | number, data: T) {
    const existing = await this.findByIndex(index, value as string);
    if (existing?.id) {
      return this.update(existing.id, data);
    }
    const record = { [index]: value, ...data };
    return this.insert(record);
  }

  assertValid(data: T) {
    const validate = ajv.compile(this.def.schema);
    const valid = validate(data);
    if (!valid) {
      throw new Error(
        `Invalid data: ${validate.errors?.map(
          (e) => `(${e.instancePath}: ${e.message})`
        )}`
      );
    }
  }

  isValid(data: T) {
    const validate = ajv.compile(this.def.schema);
    return validate(data);
  }
}
