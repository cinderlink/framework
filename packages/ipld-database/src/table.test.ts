import { TestDIDDag } from "@candor/test-adapters/src/dag";
import { createSeed, createDID } from "../../client";
import type { DID } from "dids";
import { describe, it, expect, beforeEach } from "vitest";
import { Table } from "./table";
import { BlockData, TableDefinition } from "@candor/core-types";

const validDefinition: TableDefinition = {
  schemaId: "test",
  encrypted: false,
  indexes: {
    name: {
      unique: true,
      fields: ["name"],
    },
  },
  aggregate: {
    count: "max",
  },
  schema: {
    type: "object",
    properties: {
      name: { type: "string" },
      count: { type: "number" },
    },
  },
  searchOptions: {
    fields: ["name", "id"],
  },
  rollup: 10,
};

let seed: Uint8Array;
let dag: TestDIDDag;
let did: DID;
describe("@candor/ipld-database/table", () => {
  beforeEach(async () => {
    seed = await createSeed("test seed");
    did = await createDID(seed);
    dag = new TestDIDDag(did);
  });

  it("should create a current block", () => {
    const table = new Table("test", validDefinition, dag);
    expect(table.currentBlock).toMatchSnapshot();
  });

  it("should validate records", () => {
    const table = new Table("test", validDefinition, dag);
    expect(() => table.assertValid({ name: "foo", count: 1 })).not.toThrow();
    expect(() => table.assertValid({ name: "foo", count: "1" })).toThrow();
    expect(() => table.assertValid({ name: 1, count: 1 })).toThrow();
  });

  it("should insert records", async () => {
    const table = new Table("test", validDefinition, dag);
    const id = await table.insert({ name: "foo", count: 1 });
    expect(Object.values(table.currentBlock.cache?.records || {}).length).toBe(
      1
    );
    expect(id).toBe(1);
    const results = await table.query().select().where("id", "=", id).execute();
    expect(results).toMatchInlineSnapshot(`
      TableQueryResult {
        "rows": [
          {
            "count": 1,
            "id": 1,
            "name": "foo",
          },
        ],
      }
    `);
    const record = results.first();
    expect(record).toMatchInlineSnapshot(`
      {
        "count": 1,
        "id": 1,
        "name": "foo",
      }
    `);
  });

  it("should index records", async () => {
    const table = new Table("test", validDefinition, dag);
    await table.insert({ name: "foo", count: 1 });
    expect(table.currentBlock.cache?.filters?.indexes).toMatchSnapshot();
  });

  it("should search records", async () => {
    const table = new Table("test", validDefinition, dag);
    await table.insert({ name: "foo", count: 1 });
    await table.insert({ name: "bar", count: 1 });
    await table.insert({ name: "baz", count: 1 });
    const results = await table.search("foo");
    expect(results).toMatchInlineSnapshot("[]");
  });

  it("should rollup records", async () => {
    const table = new Table("test", validDefinition, dag);
    for (let i = 0; i < 11; i++) {
      await table.insert({ name: `test #${i}`, count: i });
    }
    expect(Object.values(table.currentBlock.cache?.records || {}).length).toBe(
      1
    );
    expect(table.currentBlock.cache?.prevCID).not.toBeUndefined();
  });

  it("should aggregate records", async () => {
    const table = new Table("test", validDefinition, dag);
    for (let i = 0; i < 11; i++) {
      await table.insert({ name: `test #${i}`, count: i });
    }
    const prevBlock = await dag.load<BlockData>(
      table.currentBlock.cache!.prevCID! as any
    );
    expect(prevBlock?.filters.aggregates).toMatchInlineSnapshot(`
      {
        "count": 9,
      }
    `);
  });

  it("should rollup records with indexes", async () => {
    const table = new Table("test", validDefinition, dag);
    for (let i = 0; i < 11; i++) {
      await table.insert({ name: `test #${i}`, count: i });
    }
    const prevBlock = await dag.load<BlockData>(
      table.currentBlock.cache!.prevCID! as any
    );
    expect(prevBlock?.filters.indexes).toMatchSnapshot();
  });

  it("should rollup records with aggregates", async () => {
    const table = new Table("test", validDefinition, dag);
    for (let i = 0; i < 11; i++) {
      await table.insert({ name: `test #${i}`, count: i });
    }
    const prevBlock = await dag.load<BlockData>(
      table.currentBlock.cache!.prevCID! as any
    );
    expect(prevBlock?.filters.aggregates).toMatchInlineSnapshot(`
      {
        "count": 9,
      }
    `);
  });

  it("should rollup records with search", async () => {
    const table = new Table("test", validDefinition, dag);
    for (let i = 0; i < 11; i++) {
      await table.insert({ name: `test #${i}`, count: i });
    }
    const results = await table.search("test #3");
    expect(results).toMatchInlineSnapshot("[]");
  });

  it("should rewrite previous blocks to execute an update operation", async () => {
    const table = new Table("test", validDefinition, dag);
    for (let i = 0; i < 11; i++) {
      await table.insert({ name: `test #${i}`, count: i });
    }
    await table.update(2, { name: "test three", count: 1337 });

    const updated = await table.getById(2);
    expect(updated).toMatchInlineSnapshot(`
      {
        "count": 1337,
        "id": 2,
        "name": "test three",
      }
    `);

    const records = await (await table.query().select().execute()).all();
    expect(records).toMatchInlineSnapshot(`
      [
        {
          "count": 0,
          "id": 1,
          "name": "test #0",
        },
        {
          "count": 1337,
          "id": 2,
          "name": "test three",
        },
        {
          "count": 2,
          "id": 3,
          "name": "test #2",
        },
        {
          "count": 3,
          "id": 4,
          "name": "test #3",
        },
        {
          "count": 4,
          "id": 5,
          "name": "test #4",
        },
        {
          "count": 5,
          "id": 6,
          "name": "test #5",
        },
        {
          "count": 6,
          "id": 7,
          "name": "test #6",
        },
        {
          "count": 7,
          "id": 8,
          "name": "test #7",
        },
        {
          "count": 8,
          "id": 9,
          "name": "test #8",
        },
        {
          "count": 9,
          "id": 10,
          "name": "test #9",
        },
        {
          "count": 10,
          "id": 11,
          "name": "test #10",
        },
      ]
    `);
  });

  describe("upsert", () => {
    it("should update inserted records", async () => {
      const table = new Table("test", validDefinition, dag);
      for (let i = 0; i < 11; i++) {
        await table.insert({ name: `test #${i}`, count: i });
      }
      const result = await table.getById(2);
      expect(result).toMatchObject({
        id: 2,
        name: "test #1",
        count: 1,
      });
      await table.upsert("id", result?.id, {
        name: "test three",
        count: 1337,
      });

      const updated = await table.getById(result?.id as number);
      expect(updated).toMatchObject({
        id: result?.id,
        name: "test three",
        count: 1337,
      });
    });

    it("should insert new records", async () => {
      const table = new Table("test", validDefinition, dag);
      for (let i = 0; i < 11; i++) {
        await table.upsert("name", `test #${i}`, { count: i });
      }
      const records = await table.query().select().execute();
      expect(records.all().length).toBe(11);
      const result = await table.getById(2);
      expect(result).toMatchObject({
        id: 2,
        name: "test #1",
        count: 1,
      });
    });
  });
});
