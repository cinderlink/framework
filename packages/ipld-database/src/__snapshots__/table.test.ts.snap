// Vitest Snapshot v1

exports[`@candor/ipld-database/table > should create a current block 1`] = `
{
  "headers": {
    "encrypted": false,
    "index": 0,
    "recordsFrom": 1,
    "recordsTo": 1,
    "schema": "test",
    "table": "test",
  },
  "prevCID": undefined,
}
`;

exports[`@candor/ipld-database/table > should index records 1`] = `
{
  "name": {
    "z4TJrksEjp8": {
      "ids": [
        1,
      ],
      "values": [
        "foo",
      ],
    },
  },
}
`;

exports[`@candor/ipld-database/table > should rollup records with indexes 1`] = `
{
  "name": {
    "zPbkLwYJQbRmZQwe": {
      "ids": [
        1,
      ],
      "values": [
        "test #0",
      ],
    },
    "zPbkLwYJQbRmZkRa": {
      "ids": [
        2,
      ],
      "values": [
        "test #1",
      ],
    },
    "zPbkLwYJQbRma5uW": {
      "ids": [
        3,
      ],
      "values": [
        "test #2",
      ],
    },
    "zPbkLwYJQbRmaRPS": {
      "ids": [
        4,
      ],
      "values": [
        "test #3",
      ],
    },
    "zPbkLwYJQbRmaksN": {
      "ids": [
        5,
      ],
      "values": [
        "test #4",
      ],
    },
    "zPbkLwYJQbRmb6MJ": {
      "ids": [
        6,
      ],
      "values": [
        "test #5",
      ],
    },
    "zPbkLwYJQbRmbRqE": {
      "ids": [
        7,
      ],
      "values": [
        "test #6",
      ],
    },
    "zPbkLwYJQbRmbmKA": {
      "ids": [
        8,
      ],
      "values": [
        "test #7",
      ],
    },
    "zPbkLwYJQbRmc6o6": {
      "ids": [
        9,
      ],
      "values": [
        "test #8",
      ],
    },
    "zPbkLwYJQbRmcSH2": {
      "ids": [
        10,
      ],
      "values": [
        "test #9",
      ],
    },
  },
}
`;
